File: Examples/test-suite/perl5/access_change_runme.pl
use strict;
use warnings;
use Test::More tests => 26;
BEGIN { use_ok 'access_change' }
require_ok 'access_change';

my $k = access_change::BaseInt->new();

eval { $k->PublicProtectedPublic1() };
is($@, '', 'BaseInt.PublicProtectedPublic1()');
eval { $k->PublicProtectedPublic2() };
is($@, '', 'BaseInt.PublicProtectedPublic2()');
eval { $k->PublicProtectedPublic3() };
is($@, '', 'BaseInt.PublicProtectedPublic3()');
eval { $k->PublicProtectedPublic4() };
is($@, '', 'BaseInt.PublicProtectedPublic4()');
eval { $k->WasProtected1() };
isnt($@, '', 'BaseInt.WasProtected1()');
eval { $k->WasProtected2() };
isnt($@, '', 'BaseInt.WasProtected2()');
eval { $k->WasProtected3() };
isnt($@, '', 'BaseInt.WasProtected3()');
eval { $k->WasProtected4() };
isnt($@, '', 'BaseInt.WasProtected4()');

$k = access_change::DerivedInt->new();

TODO: {
    local $TODO = '"using" support may be botched';
    eval { $k->PublicProtectedPublic1() };
    isnt($@, '', 'DerivedInt.PublicProtectedPublic1()');
    eval { $k->PublicProtectedPublic2() };
    isnt($@, '', 'DerivedInt.PublicProtectedPublic2()');
}
eval { $k->PublicProtectedPublic3() };
is($@, '', 'DerivedInt.PublicProtectedPublic3()');
eval { $k->PublicProtectedPublic4() };
is($@, '', 'DerivedInt.PublicProtectedPublic4()');
eval { $k->WasProtected1() };
is($@, '', 'DerivedInt.WasProtected1()');
eval { $k->WasProtected2() };
is($@, '', 'DerivedInt.WasProtected2()');
eval { $k->WasProtected3() };
is($@, '', 'DerivedInt.WasProtected3()');
eval { $k->WasProtected4() };
is($@, '', 'DerivedInt.WasProtected4()');

$k = access_change::BottomInt->new();

eval { $k->PublicProtectedPublic1() };
is($@, '', 'BaseInt.PublicProtectedPublic1()');
eval { $k->PublicProtectedPublic2() };
is($@, '', 'BaseInt.PublicProtectedPublic2()');
eval { $k->PublicProtectedPublic3() };
is($@, '', 'BaseInt.PublicProtectedPublic3()');
eval { $k->PublicProtectedPublic4() };
is($@, '', 'BaseInt.PublicProtectedPublic4()');
eval { $k->WasProtected1() };
is($@, '', 'BaseInt.WasProtected1()');
eval { $k->WasProtected2() };
is($@, '', 'BaseInt.WasProtected2()');
eval { $k->WasProtected3() };
TODO: {
    local $TODO = '"using" support may be botched';
    isnt($@, '', 'BaseInt.WasProtected3()');
    eval { $k->WasProtected4() };
    isnt($@, '', 'BaseInt.WasProtected4()');
}

File: Examples/test-suite/lua/access_change_runme.lua
local v=require("access_change")

local baseInt = v.BaseInt()
assert(baseInt ~= nil)
assert(baseInt:PublicProtectedPublic1() == nil)
assert(baseInt:PublicProtectedPublic2() == nil)
assert(baseInt:PublicProtectedPublic3() == nil)
assert(baseInt:PublicProtectedPublic4() == nil)
-- Protected methods
assert(not pcall(function () baseInt:WasProtected1() end))
assert(not pcall(function () baseInt:WasProtected2() end))
assert(not pcall(function () baseInt:WasProtected3() end))
assert(not pcall(function () baseInt:WasProtected4() end))

local derivedInt = v.DerivedInt()
assert(derivedInt ~= nil)
-- Here there are public
assert(derivedInt:WasProtected1() == nil)
assert(derivedInt:WasProtected2() == nil)
assert(derivedInt:WasProtected3() == nil)
assert(derivedInt:WasProtected4() == nil)
-- inherite public from base
assert(derivedInt:PublicProtectedPublic1() == nil)
assert(derivedInt:PublicProtectedPublic2() == nil)
assert(derivedInt:PublicProtectedPublic3() == nil)
assert(derivedInt:PublicProtectedPublic4() == nil)

local bottomInt = v.BottomInt()
assert(bottomInt ~= nil)
-- All methods are public
assert(bottomInt:WasProtected1() == nil)
assert(bottomInt:WasProtected2() == nil)
assert(bottomInt:WasProtected3() == nil)
assert(bottomInt:WasProtected4() == nil)
assert(bottomInt:PublicProtectedPublic1() == nil)
assert(bottomInt:PublicProtectedPublic2() == nil)
assert(bottomInt:PublicProtectedPublic3() == nil)
assert(bottomInt:PublicProtectedPublic4() == nil)
